hazelcast/src/main/java/com/hazelcast/query/impl/CompositeValue.java

```diff
@@ -162,7 +162,14 @@ public boolean equals(Object o) {
         }
 
         CompositeValue that = (CompositeValue) o;
-        return Arrays.equals(components, that.components);
+        assert components.length == that.components.length;
+
+        for (int i = 0; i < components.length; ++i) {
+            if (!Comparables.equal(components[i], that.components[i])) {
```

Comment:
Do we need to adapt the behaviour of `CompositeValue#hashCode` because of this? In another words, can it happen that `compareTo` returns `0` but `equals` returns `false`?

Reply:
That's really a good catch! Looks like there is no need in using `Comparables.equal` here since composite values are always canonicalized externally before calling `equals/hashCode`. I will double check that.